unison-cli/src/Unison/Share/Sync.hs

```diff
@@ -606,64 +619,141 @@ getCausalHashByPath httpClient unisonShareUrl repoPath =
-      -- Each request only contains a certain maximum number of entities; split the set of hashes we need to upload into
-      -- those we will upload right now, and those we will begin uploading on the next iteration of the loop.
-      let (hashesSet, nextHashes) =
-            case NESet.splitAt 50 allHashesSet of
-              This hs1 -> (hs1, Set.empty)
-              That hs2 -> (hs2, Set.empty) -- impossible, this only happens if we split at 0
-              These hs1 hs2 -> (hs1, NESet.toSet hs2)
-
-      let hashesList = NESet.toAscList hashesSet
-      -- Get each entity that the server is missing out of the database.
-      entities <- Sqlite.runTransaction conn (traverse expectEntity hashesList)
-
-      let uploadEntities :: IO Share.UploadEntitiesResponse
-          uploadEntities = do
-            -- Timing.time ("uploadEntities with " <> show (NESet.size hashesSet) <> " hashes.") do
-            httpUploadEntities
-              httpClient
-              unisonShareUrl
-              Share.UploadEntitiesRequest
-                { entities = NEMap.fromAscList (List.NonEmpty.zip hashesList entities),
-                  repoName
-                }
-
-      -- The new upload count *if* we make a successful upload.
-      let newUploadCount = uploadCount + NESet.size hashesSet
-
-      uploadEntities >>= \case
-        Share.UploadEntitiesNeedDependencies (Share.NeedDependencies moreHashes) -> do
-          let newAllHashesSet = union10 moreHashes nextHashes
-          uploadProgressCallback newUploadCount (NESet.size newAllHashesSet)
-          loop newUploadCount newAllHashesSet
-        Share.UploadEntitiesNoWritePermission _ -> pure False
-        Share.UploadEntitiesHashMismatchForEntity {} -> pure False
-        Share.UploadEntitiesSuccess -> do
-          case NESet.nonEmptySet nextHashes of
-            Nothing -> do
-              uploadProgressCallback newUploadCount 0
-              pure True
-            Just nextHashes1 -> do
-              uploadProgressCallback newUploadCount (NESet.size nextHashes1)
-              loop newUploadCount nextHashes1
-
--- Union a non-empty set and a set.
-union10 :: Ord a => NESet a -> Set a -> NESet a
-union10 xs ys =
-  case NESet.nonEmptySet ys of
-    Nothing -> xs
-    Just zs -> NESet.union xs zs
+    dispatcher ::
+      Ki.Scope ->
+      TVar (Set Hash32) ->
+      TVar (Set Hash32) ->
+      TVar Int ->
+      TVar (Set Int) ->
+      TMVar () ->
+      IO Bool
+    dispatcher scope hashesVar dedupeVar nextWorkerIdVar workersVar workerFailedVar = do
+      loop
+      where
+        loop :: IO Bool
+        loop =
+          doJob [checkForFailureMode, dispatchWorkMode, checkIfDoneMode]
+
+        doJob :: [STM UploadDispatcherJob] -> IO Bool
+        doJob jobs =
+          atomically (asum jobs) >>= \case
+            UploadDispatcherReturnFailure -> pure False
+            UploadDispatcherForkWorkerWhenAvailable hashes -> doJob [forkWorkerMode hashes, checkForFailureMode]
```

Comment:
Hrmm, was it intentional to check for failure only if we fail to start a new job?

I'm confused about the semantics of checking for failure, it seems in some spots it bails, and spots like here it bails only conditionally.